### ISSUE TITLE
Autocomplete search problematic

### ISSUE THREAD

**Issue by DeathBorn** [sehmaschine/django-grappelli#706]
Autocomplete search is a bit problematic and it should be changed to something better
[Code](https://github.com/sehmaschine/django-grappelli/blob/master/grappelli/views/related.py#L149)

``` python
....
def get_searched_queryset(self, qs):
        model = self.model
        term = self.GET["term"]

        try:
            term = model.autocomplete_term_adjust(term)
        except AttributeError:
            pass

        search_fields = get_autocomplete_search_fields(self.model)
        if search_fields:
            for word in term.split():   # Here is the problem
                search = [models.Q(**{smart_text(item): smart_text(word)}) for item in search_fields]
                search_qs = QuerySet(model)
                search_qs.query.select_related = qs.query.select_related
                search_qs = search_qs.filter(reduce(operator.or_, search))
                qs &= search_qs
        else:
            qs = model.objects.none()
        return qs
.....
```

This place in code splits search term by space in multiple search terms. It seems nice and sound but when you need to search by term consisting of more than one word combination we have a problem. It would be better to create combinations too.
For example

``` python
stuff = term.split()    #["a","b","c", "d"]
for i, j in itertools.combinations(range(len(stuff) + 1), 2):
    print(" ".join(stuff[i:j]))

a
a b
a b c
a b c d
b
b c
b c d
c
c d
d
```

But now I need to make additional field in database with spaces replaced with "-" and use adjust autocomplete term which is not nice.

**Comment by sehmaschine** [comment]
sounds good. can you provied tests for this behaviour?

**Comment by BertrandBordage** [comment]
@DeathBorn: This is a really tricky issue… That’s why search engines are for. You see the problem with the '-'? There’s plenty of other characters that need to be dealt with, like the many types of spaces, 'ß' vs 'ss' etc. 

If we implement what you describe, we’ll end up with an enormous query: it needs to filter against all the combinations of splitted words for all the search fields. And the query will contain dozens of `OR`, which is bad for performance.
Suppose the search fields are `first_name__icontains`, `middle_name__icontains` & `last_name__icontains`, and you search `Antonio Lucio Vivaldi`. This is a simple case, I have [far more complex cases](https://github.com/dezede/dezede/blob/002034e54ff8e10401fb6015f4d91660f61b3d72/libretto/models/oeuvre.py#L957-L971). The query will look like that:
```python
Composer.objects.filter(
    (Q(first_name__icontains='Antonio')
     & Q(middle_name__icontains='Lucio')
     & Q(last_name__icontains='Vivaldi'))
    | (Q(first_name__icontains='Antonio')
       & Q(last_name__icontains='Lucio')
       & Q(middle_name__icontains='Vivaldi'))
    | (Q(first_name__icontains='Antonio')
       & Q(middle_name__icontains='Lucio Vivaldi'))
    | (Q(first_name__icontains='Antonio')
       & Q(last_name__icontains='Lucio Vivaldi'))
    | (Q(first_name__icontains='Antonio Lucio')
       & Q(middle_name__icontains='Vivaldi'))
    | (Q(first_name__icontains='Antonio Lucio')
       & Q(last_name__icontains='Vivaldi'))
    | Q(first_name__icontains='Antonio Lucio Vivaldi')
    | (Q(middle_name__icontains='Antonio')
       & Q(first_name__icontains='Lucio')
       & Q(last_name__icontains='Vivaldi'))
    | (Q(middle_name__icontains='Antonio')
       & Q(last_name__icontains='Lucio')
       & Q(first_name__icontains='Vivaldi'))
    | (Q(middle_name__icontains='Antonio')
       & Q(first_name__icontains='Lucio Vivaldi'))
    | (Q(middle_name__icontains='Antonio')
       & Q(last_name__icontains='Lucio Vivaldi'))
    | (Q(middle_name__icontains='Antonio Lucio')
       & Q(first_name__icontains='Vivaldi'))
    | (Q(middle_name__icontains='Antonio Lucio')
       & Q(last_name__icontains='Vivaldi'))
    | Q(middle_name__icontains='Antonio Lucio Vivaldi')
    | (Q(last_name__icontains='Antonio')
       & Q(first_name__icontains='Lucio')
       & Q(middle_name__icontains='Vivaldi'))
    | (Q(last_name__icontains='Antonio')
       & Q(middle_name__icontains='Lucio')
       & Q(first_name__icontains='Vivaldi'))
    | (Q(last_name__icontains='Antonio')
       & Q(first_name__icontains='Lucio Vivaldi'))
    | (Q(last_name__icontains='Antonio')
       & Q(middle_name__icontains='Lucio Vivaldi'))
    | (Q(last_name__icontains='Antonio Lucio')
       & Q(first_name__icontains='Vivaldi'))
    | (Q(last_name__icontains='Antonio Lucio')
       & Q(middle_name__icontains='Vivaldi'))
    | Q(last_name__icontains='Antonio Lucio Vivaldi')
)
```

And this is only a simple case, imagine when users search 5 words in a 10 search fields scenario… I tested the query above against a PostgreSQL table of 9000 people where none match the query: takes 60 ms to complete, and there are indexes on these columns. In comparison, the current query django-grappelli executes is the following and takes 24 ms to run on the same table:

```python
Composer.objects.filter(
    (Q(first_name__icontains='Antonio')
     | Q(middle_name__icontains='Antonio')
     | Q(last_name__icontains='Antonio'))
    & (Q(first_name__icontains='Lucio')
       | Q(middle_name__icontains='Lucio')
       | Q(last_name__icontains='Lucio'))
    & (Q(first_name__icontains='Vivaldi')
       | Q(middle_name__icontains='Vivaldi')
       | Q(last_name__icontains='Vivaldi')))
```

The good news is that this approach would solve #585. But it will not sort the results by relevance, which can only be done by a search engine. And of course, performance with a search engine would be unmatched.

To sum up, **it would partially improve autocompletion while decreasing performance on most cases**.

In my opinion, what we probably need to do first is allow users to specify their own `AutocompleteLookup` subclass in the settings. Since a lot of sites have full-text search, using haystack for example, this would be really useful.

**Comment by DeathBorn** [comment]
Your point is very good.

**Comment by adamchainz** [comment]
This breaks sometimes for us sometimes, we get `OperationalError: (1116, 'Too many tables; MariaDB can only use 61 tables in a join')` when a search query with many words is used and the filters join tables

**Comment by sehmaschine** [comment]
Closing this since we plan to remove Grappelli autocompletes in favour of the now available Django autocompletes.
